runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -796,71 +797,87 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
          isStoreValueVT = isValue(storeValueConstraint);
          }
 
-      // If the array's component type is definitely not a value type, or if the value
-      // being assigned in an array store operation is definitely not a value type, add
-      // a delayed transformation to replace the helper call with inline code to
-      // perform the array element access.
-      //
-      if ((arrayConstraint != NULL && isCompTypeVT == TR_no)
-          || (isStoreFlattenableArrayElement && isStoreValueVT == TR_no))
+      if (arrayConstraint &&
+          isArrayElementFlattened(arrayConstraint) == TR_yes &&
+          fe()->getComponentClassFromArrayClass(arrayConstraint->getClass()))
```

Comment:
For the array store operation, I think this needs to confirm that the type of the value is the same as the component type of the array - it could be storing a `java/lang/Object` or a value whose type is not known.  Those would require an `ArrayStoreException` to be thrown if the type did not match the component type of the array, or a `NullPointerException` if the value turned out to be a null reference.  I think calling the helper would be the easiest way to enforce that checking for the unlikely case where the JIT knows the array is a value type array but doesn't know the type of the value that's being stored.

Reply:
>calling the helper would be the easiest way to enforce that checking for the unlikely case where the JIT knows the array is a value type array but doesn't know the type of the value that's being stored.

Good catch! Is there an API I could call to check if the value that's being stored is the same type as the array component class? I could check the following criteria but it doesn't guarantee matching the types between the value and the array component. 

```
if (!storeValueConstraint->isNonNullObject() || !isValue(storeValueConstraint)) 
     do not transform the helper call 
```

Reply:
If you take a look at [constrainArrayStoreChk](https://github.com/eclipse/omr/blob/b5643d019207e22e3a6bcb49d09069200289b639/compiler/optimizer/VPHandlers.cpp#L12244-L12311) in VPHandlers.cpp in OMR, there is code that performs that same kind of checking.  Perhaps that code could be made common between the two - I haven't looked at it carefully, so I'm not sure whether there is some checking that would inappropriate or irrelevant for value type arrays.

Reply:
Thanks for the example @hzongaro! After looking at `constrainArrayStoreChk`,  I think I'll handle the following two array element store cases differently: 

1. If the value could be `NULL` (`!isNonNull()`) during the compile time, insert a `NULLCHK` on the value node before the transformation in `transformFlattenedArrayElementStore`.  This way more store calls are transformed to use sym refs if the value ends up being non-null at runtime. 

2. If the type of the value is not the same as the component type of the array (by checking `isInstanceOf`) during the compile time, do not transform the store helper call.

Reply:
Addressed in e5456bc